Java/core/src/main/java/com/amazon/randomcutforest/preprocessor/ImputePreprocessor.java

```diff
@@ -381,54 +257,59 @@ protected int determineGap(long timestampGap, double averageGap) {
      *
-     * @param descriptor   description of the current point
      * @param averageGap   the gap in timestamps
      * @param changeForest boolean determining if we commit to the forest or not
      * @param forest       the resident RCF
      * @return the next shingle
      */
-    protected double[] generateShingle(AnomalyDescriptor descriptor, double averageGap, boolean changeForest,
-            RandomCutForest forest) {
-        double[] input = descriptor.getCurrentInput();
-        long timestamp = descriptor.getInputTimestamp();
+    protected float[] generateShingle(double[] inputTuple, long timestamp, int[] missingValues, double averageGap,
+            boolean changeForest, RandomCutForest forest) {
         long lastInputTimeStamp = previousTimeStamps[shingleSize - 1];
-        int[] missingValues = descriptor.getMissingValues();
-        checkArgument(missingValues == null || (imputationMethod != LINEAR && imputationMethod != NEXT),
-                " cannot perform imputation on most recent missing value with this method");
-        /*
-         * Note only ZERO, FIXED_VALUES, PREVIOUS and RCF are reasonable options if
-         * missing values are present.
-         */
-
-        checkArgument(internalTimeStamp > 0, "imputation should have forced normalization");
-        double[] savedInput = getShingledInput(shingleSize - 1);
-
-        // previous value should be defined
-        double[] previous = new double[inputLength];
-        System.arraycopy(lastShingledInput, lastShingledInput.length - inputLength, previous, 0, inputLength);
+        double[] input = Arrays.copyOf(inputTuple, inputLength);
+        double[] previous = getShingledInput(shingleSize - 1);
+
+        if (missingValues != null && imputationMethod != RCF) {
+            // for STREAMING_IMPUTE the timestamp cannot be missing
+            // hence missingValues[] can be 0 to inputLength - 1
+            // for next and Linear there are no current values
+            // we are forced to use fixedvalues or previous
+            for (int missingValue : missingValues) {
+                input[missingValue] = (defaultFill == null) ? previous[missingValue] : defaultFill[missingValue];
+            }
+        }
+
+        double[] savedInput = Arrays.copyOf(previous, inputLength);
         // using the global dependency
         int numberToImpute = determineGap(timestamp - lastInputTimeStamp, averageGap) - 1;
         if (numberToImpute > 0) {
-            descriptor.setNumberOfNewImputes(numberToImpute);
             double step = 1.0 / (numberToImpute + 1);
             // the last impute corresponds to the current observed value
             for (int i = 0; i < numberToImpute; i++) {
-                double[] result = impute(false, descriptor, step * (i + 1), previous, forest);
+                // only the last tuple is partial
+                double[] result = impute(input, missingValues, false, step * (i + 1), previous, forest);
```

Comment:
You are imputing for all of the gaps. If the number of gaps are huge, this will take a lot of time. Anyway to prevent that?

Reply:
Good q. It would definitely take a long time if we use RCF (and that is indeed better). Maybe we should have alternate logic if the gap is very large. (Also if the gap is very large then we are restarting froma completely fake shingle ...). Let me get back on this.

Reply:
Done. But let's move any more changes to impute processor to subsequent PRs. This new refactor is already much better than the 3.8 version (and has much higher test coverage).